pydm/data_store/__init__.py

```diff
@@ -0,0 +1,182 @@
+    """
+    Enum class which holds the keys expected by the PyDMWidgets when used with
+    a structured data to parse the data in search for the needed fields.
+    """
+    CONNECTION = 'CONNECTION'
+    VALUE = 'VALUE'
+    SEVERITY = 'SEVERITY'
+    WRITE_ACCESS = 'WRITE_ACCESS'
+    ENUM_STRINGS = 'ENUM_STRINGS'
+    UNIT = 'UNIT'
+    PRECISION = 'PRECISION'
+    UPPER_LIMIT = 'UPPER_LIMIT'
+    LOWER_LIMIT = 'LOWER_LIMIT'
+
+    @staticmethod
+    def generate_introspection_for(connection_key=None, value_key=None,
+                                   severity_key=None, write_access_key=None,
+                                   enum_strings_key=None, unit_key=None,
+                                   precision_key=None, upper_limit_key=None,
+                                   lower_limit_key=None
+                                   ):
+        """
+        Generates an introspection dictionary for a given set of keys.
+        This is used by PyDMWidgets to map the needed keys in a structured
+        data source into the fields needed.
+
+        Parameters
+        ----------
+        connection_key : str
+            The key for the connection status information at the data
+            dictionary
+        value_key : str
+            The key for the value information at the data dictionary
+        severity_key : str
+            The key for the severity information at the data dictionary
+        write_access_key : str
+            The key for the write access information at the data dictionary
+        enum_strings_key : str
+            The key for the enum strings information at the data dictionary
+        unit_key : str
+            The key for the engineering unit information at the data dictionary
+        precision_key : str
+            The key for the precision information at the data dictionary
+        upper_limit_key : str
+            The key for the upper limit information at the data dictionary
+        lower_limit_key : str
+            The key for the lower limit information at the data dictionary
+
+        Returns
+        -------
+        introspection : dict
+
+        """
+        lookup_table = [
+            (connection_key, DataKeys.CONNECTION),
+            (value_key, DataKeys.VALUE),
+            (severity_key, DataKeys.SEVERITY),
+            (write_access_key, DataKeys.WRITE_ACCESS),
+            (enum_strings_key, DataKeys.ENUM_STRINGS),
+            (unit_key, DataKeys.UNIT),
+            (precision_key, DataKeys.PRECISION),
+            (upper_limit_key, DataKeys.UPPER_LIMIT),
+            (lower_limit_key, DataKeys.LOWER_LIMIT)
+        ]
+        introspection = dict()
+
+        for val, key in lookup_table:
+            if val:
+                introspection[key] = val
+
+        return introspection
+
+
+DEFAULT_INTROSPECTION = {k: v for k, v in DataKeys.__dict__.items()
+                         if isinstance(v, str) and not k.startswith('_')}
+
+
+class DataStore(object):
```

Comment:
Might consider instantiating this once, and referring to it later.

e.g.,

```python
class _DataStore:
    ...

DataStore = _DataStore()
```

Reply:
Since it is a singleton, would it make a difference?

Reply:
It would make a bit of a difference as you would not have to have strange `__new__` methods and such.

Reply:
Addressed with 15403f7b